File: F - Devsinc Flood/sol.py
def input_data():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    n = int(data[0])
    lst = list(map(int, data[1:]))
    
    a = [0]
    for x in lst:
        if len(a) == 0 or x != a[-1]:
            a.append(x)
    
    n = len(a) - 1
    return n, a

def solve(n, a):
    prev_dp = [0] * (n + 1)
    
    for i in range(1, n + 1):
        curr_dp = [0] * (n + 1)
        for j in range(1, n + 1):
            if a[i] == a[n - j + 1]:
                curr_dp[j] = prev_dp[j-1] + 1
            else:
                curr_dp[j] = max(prev_dp[j], curr_dp[j-1])
        prev_dp = curr_dp
    
    return prev_dp

def output(n, dp):
    print(n - (dp[n] + 1) // 2)

if __name__ == "__main__":
    n, a = input_data()
    dp = solve(n, a)
    output(n, dp)

File: F - Devsinc Flood/sol.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int maxN = 5008;

int n;
int dp[maxN][maxN];
vector<int> a(1), b;
vector<int> ans;

void input() {
	cin >> n;
	for (int i = 0; i < n; i++) {
		int x; cin >> x;
		if (x != a.back()) a.push_back(x);
	}
	n = a.size() - 1;

	b = a;
	reverse(b.begin() + 1, b.end());
}

void solve() {
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			if (a[i] == b[j]) {dp[i][j] = dp[i-1][j-1] + 1;}
			else {dp[i][j] = max(dp[i-1][j], dp[i][j-1]);}
		}
	}
}

void output() {
	cout << n - (dp[n][n] + 1)/2 << '\n';
}

int main() {
	
	input();
	solve();
	output();
	
	return 0;
}

